clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StorePartitionDataReceiver.java

```diff
@@ -75,12 +123,1462 @@ public void write(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> cons
+        long beforeProcessingPerRecordTimestampNs = System.nanoTime();
+        List<PubSubMessageProcessedResultWrapper<KafkaKey, KafkaMessageEnvelope, Long>> processedResults =
+            ingestionBatchProcessor.process(
+                batch,
+                partitionConsumptionState,
+                topicPartition.getPartitionNumber(),
+                kafkaUrl,
+                kafkaClusterId,
+                beforeProcessingPerRecordTimestampNs,
+                beforeProcessingBatchRecordsTimestampMs);
+
+        for (PubSubMessageProcessedResultWrapper<KafkaKey, KafkaMessageEnvelope, Long> processedRecord: processedResults) {
+          totalBytesRead += handleSingleMessage(
+              processedRecord,
+              topicPartition,
+              partitionConsumptionState,
+              kafkaUrl,
+              kafkaClusterId,
+              beforeProcessingPerRecordTimestampNs,
+              beforeProcessingBatchRecordsTimestampMs,
+              metricsEnabled,
+              elapsedTimeForPuttingIntoQueue);
+        }
+      } finally {
+        ingestionBatchProcessor.unlockKeys(keyLockMap);
+      }
+    }
+
+    updateMetricsAndEnforceQuota(
+        totalBytesRead,
+        topicPartition.getPartitionNumber(),
+        elapsedTimeForPuttingIntoQueue,
+        beforeProcessingBatchRecordsTimestampMs);
+  }
+
+  private void updateMetricsAndEnforceQuota(
+      long totalBytesRead,
+      int partition,
+      ValueHolder<Double> elapsedTimeForPuttingIntoQueue,
+      long beforeProcessingBatchRecordsTimestampMs) {
+    /**
+     * Even if the records list is empty, we still need to check quota to potentially resume partition
+     */
+    final StorageUtilizationManager storageUtilizationManager = storeIngestionTask.getStorageUtilizationManager();
+    storageUtilizationManager.enforcePartitionQuota(partition, totalBytesRead);
+
+    if (storeIngestionTask.isMetricsEmissionEnabled()) {
+      HostLevelIngestionStats hostLevelIngestionStats = storeIngestionTask.getHostLevelIngestionStats();
+      if (totalBytesRead > 0) {
+        hostLevelIngestionStats.recordTotalBytesReadFromKafkaAsUncompressedSize(totalBytesRead);
+      }
+      if (elapsedTimeForPuttingIntoQueue.getValue() > 0) {
+        hostLevelIngestionStats.recordConsumerRecordsQueuePutLatency(
+            elapsedTimeForPuttingIntoQueue.getValue(),
+            beforeProcessingBatchRecordsTimestampMs);
+      }
+
+      hostLevelIngestionStats.recordStorageQuotaUsed(storageUtilizationManager.getDiskQuotaUsage());
+    }
+  }
+
+  public Iterable<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> validateAndFilterOutDuplicateMessagesFromLeaderTopic(
+      Iterable<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records,
+      String kafkaUrl,
+      PubSubTopicPartition topicPartition) {
+    PartitionConsumptionState partitionConsumptionState =
+        storeIngestionTask.getPartitionConsumptionState(topicPartition.getPartitionNumber());
```

Comment:
I think it'd be great if the `PCS` became a class property of the `StorePartitionDataReceiver`, rather than be held in a map inside the `SIT`. I suspect this is easier said than done, and no need to do it just yet... A large refactoring like this is ideally kept free of functional changes, and that one may be too risky. Better do it as a smaller follow up change. But just wanted to share this crazy dream 😂